### DISCUSSION TITLE
-fx is applied twice

### DISCUSSION THREAD

**Discussion by prettygoodman81** [ImageMagick/ImageMagick#6518] [Help] [answered]
### ImageMagick version

7.1.1-14

### Operating system, version and so on

macOS 12.6.7

### Description

Hello!

I try
```sh
magick mogrify -monitor -fx '{expression}' image1.png image2.png ...
```
and get outputs
```sh
Load/Image/{path}[image1.png]: 63 of 64, 100% complete
FxNew/Image/{path}[image1.png]: 63 of 64, 100% complete
FxNew/Image/{path}[image1.png]: 63 of 64, 100% complete
Save/Image/{path}[image1.png]: 63 of 64, 100% complete
...
```
Why is the `-fx` applied twice?
Is there a way to apply the `-fx` only once?

**Comment by snibgo** [comment]
In your example, `-fx` is executed four times, twice per input image. I think this is a bug.

in `MagickWand/mogrify.c`, `FxImage()` is called by functions `MogrifyImage()` and `MogrifyImageList()`. This seems wrong to me. I think `FxImage()` should be called by only one of these.

Please give a full example, showing what expression you give to `-fx`. In particular, I would like to know if the expression uses data from one or more images.

**Comment by snibgo** [comment]
Another thought: the following command ...
```
magick mogrify -monitor -fx "v/2" -format jpg *.png
```
... fails with the error:
```
mogrify: Symbol 'v' but fewer than two images at 'v/2' @ error/fx.c/GetOperand/2235.
```
This shows that `-fx` is being called with only one image in the current image list. This seems reasonable to me. 

```
magick mogrify -monitor -fx "u/2" -format jpg *.png

load image[r.png]: 45 of 46, 100% complete
FxNew/Image[r.jpg]: 45 of 46, 100% complete
FxNew/Image[r.jpg]: 45 of 46, 100% complete
save image[r.jpg]: 45 of 46, 100% complete
load image[toes.png]: 232 of 233, 100% complete
FxNew/Image[toes.jpg]: 232 of 233, 100% complete
FxNew/Image[toes.jpg]: 232 of 233, 100% complete
save image[toes.jpg]: 232 of 233, 100% complete
```
The `-fx` is called twice for each image, so each image is divided by two twice. This seems wrong.

**Comment by prettygoodman81** [comment]
Thank you for your comment.

Sorry for the lack of explanation.
The expression is the algorithm for converting regular glyphs to bold glyphs:
```
if (p == white && p[1,0] == black,
    white,
    if (p[-1,0] == black,
        black,
        p
    )
)
```
I have a lot of disjointed glyph images and I want to apply the formula to each glyph image.

**Comment by prettygoodman81** [comment]
I'm using google translate, so sorry if it's weird or my reply is late

**Reply by dlemstra** [reply]
Don't worry about a late reply or a weird message. We will ask questions if we don;t understand what you are trying to tell us. And our reply might also be late because we have other things to do.

**Comment by urban-warrior** [comment, marked as answer]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/archive/beta/ by sometime tomorrow.

**Comment by prettygoodman81** [comment]
Confirmed resolution in beta release.
Thank you! !
